plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,196 @@
+                        DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                        getResponse.getSourceAsString()
+                    );
+                listener.onResponse(new GetRuleResponse(Map.of(id, Builder.fromXContent(parser).build()), RestStatus.OK));
+            } catch (IOException e) {
+                logger.error("Error parsing rule with ID {}: {}", id, e.getMessage());
+                listener.onFailure(e);
+            }
+        } else {
+            listener.onFailure(new ResourceNotFoundException("Rule with ID " + id + " not found."));
+        }
+    }
+
+    /**
+     * Fetch all rule from system index based on attributeFilters
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    private void fetchAllRules(Map<RuleAttribute, Set<String>> attributeFilters, ActionListener<GetRuleResponse> listener) {
+        ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext().stashContext();
+        client.prepareSearch(RULE_INDEX)
+            .setQuery(QueryBuilders.matchAllQuery())
+            .setSize(MAX_RETURN_SIZE_ALLOWED_PER_GET_REQUEST)
+            .execute(ActionListener.wrap(
+                searchResponse -> {
+                    try (ThreadContext.StoredContext context = storedContext) {
+                        handleGetAllRuleResponse(searchResponse, attributeFilters, listener);
+                    }
+                },
+                e -> {
+                    try (ThreadContext.StoredContext context = storedContext) {
+                        logger.error("Failed to fetch all rules: {}", e.getMessage());
+                        listener.onFailure(e);
+                    }
+                }
+            ));
+    }
+
+    /**
+     * Process searchResponse from index and send a GetRuleResponse
+     * @param searchResponse - Response received from index
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    private void handleGetAllRuleResponse(
+        SearchResponse searchResponse,
+        Map<RuleAttribute, Set<String>> attributeFilters,
+        ActionListener<GetRuleResponse> listener
+    ) {
+        Map<String, Rule> ruleMap = Arrays.stream(searchResponse.getHits().getHits()).map(hit -> {
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                XContentParser parser = MediaTypeRegistry.JSON.xContent()
+                    .createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, hit.getSourceAsString());
+                Rule currRule = Rule.Builder.fromXContent(parser).build();
+                if (matchesFilters(currRule, attributeFilters)) {
+                    return Map.entry(hit.getId(), currRule);
+                }
```

Comment:
we should  pass these filters to the request itself so that the filtering happens at the lucene layer and we get more accurate results back rather than all the Rules from the index ?